lib/string_map.h

```diff
@@ -0,0 +1,226 @@
+            init_data_map();
+        }
+        return *this;
+    }
+    string_map &operator=(string_map &&a) = default;
+    string_map(std::initializer_list<value_type> il) { insert(il.begin(), il.end()); }
+
+    iterator begin() noexcept { return data.begin(); }
+    const_iterator begin() const noexcept { return data.begin(); }
+    iterator end() noexcept { return data.end(); }
+    const_iterator end() const noexcept { return data.end(); }
+    reverse_iterator rbegin() noexcept { return data.rbegin(); }
+    const_reverse_iterator rbegin() const noexcept { return data.rbegin(); }
+    reverse_iterator rend() noexcept { return data.rend(); }
+    const_reverse_iterator rend() const noexcept { return data.rend(); }
+    const_iterator cbegin() const noexcept { return data.cbegin(); }
+    const_iterator cend() const noexcept { return data.cend(); }
+    const_reverse_iterator crbegin() const noexcept { return data.crbegin(); }
+    const_reverse_iterator crend() const noexcept { return data.crend(); }
+
+    bool empty() const noexcept { return data.empty(); }
+    size_type size() const noexcept { return data_map.size(); }
+    size_type max_size() const noexcept { return data_map.max_size(); }
+    bool operator==(const string_map &a) const { return data == a.data; }
+    bool operator!=(const string_map &a) const { return data != a.data; }
+    void clear() {
+        data.clear();
+        data_map.clear();
+    }
+
+    // Functions above do have `std::string_view` versions. Here we are having
+    // important special case: if `a` is not something that was interned, we do
+    // not copy / intern it, we know for sure that `a` is not in the map and we
+    // do not need to perform a lookup
+    iterator find(cstring a) { return tr_iter(data_map.find(a)); }
+    iterator find(std::string_view a) {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return data.end();
+
+        return tr_iter(data_map.find(key));
+    }
+    const_iterator find(cstring a) const { return tr_iter(data_map.find(a)); }
+    const_iterator find(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return data.end();
+
+        return tr_iter(data_map.find(key));
+    }
+
+    size_type count(cstring a) const { return data_map.count(a); }
+    size_type count(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return 0;
+
+        return data_map.count(key);
+    }
+
+    bool contains(cstring a) const { return data_map.contains(a); }
+    bool contains(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return false;
+
+        return data_map.contains(key);
+    }
+
+    template <typename Key>
+    V &operator[](Key &&k) {
+        auto it = find(key_type(std::forward<Key>(k)));
+        if (it == data.end()) {
+            it = data.emplace(data.end(), k, V());
```

Comment:
```suggestion
            it = data.emplace(data.end(), std::piecewise_construct, std::forward_as_tuple(k), std::forward_as_tuple());
```

Yes, this is ugly, but it prevents creating a temporary that is then copied to the constructor.

Reply:
Yeah, nice catch. Completely forgot this here.